src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -0,0 +1,320 @@
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+using SafeCurlHandle = Interop.libcurl.SafeCurlHandle;
+using SafeCurlMultiHandle = Interop.libcurl.SafeCurlMultiHandle;
+using curlioerr = Interop.libcurl.curlioerr;
+using curliocmd = Interop.libcurl.curliocmd;
+using CURLINFO = Interop.libcurl.CURLINFO;
+using CURLcode = Interop.libcurl.CURLcode;
+using CURLMcode = Interop.libcurl.CURLMcode;
+using curl_socket_t = System.Int32;
+using CurlPoll = Interop.libcurl.CurlPoll;
+using CurlSelect = Interop.libcurl.CurlSelect;
+
+namespace System.Net.Http
+{
+    internal partial class CurlHandler
+    {
+        private const string s_httpPrefix = "HTTP/";
+
+        private static ulong CurlReceiveHeadersCallback(
+            IntPtr buffer,
+            ulong size,
+            ulong nitems,
+            IntPtr context)
+        {
+            return ExecuteCallback(buffer, size, nitems, context,
+                (ptr, totalSize, state) =>
+                {
+                    state.CancellationToken.ThrowIfCancellationRequested();
+
+                    // The callback is invoked once per header and multi-line
+                    // headers get merged into a single line
+                    string responseHeader = Marshal.PtrToStringAnsi(ptr).Trim();
+                    HttpResponseMessage response = state.ResponseMessage;
+
+                    // TODO: Understand scenarios where multiple sets of headers are received
+                    if (!TryParseStatusLine(response, responseHeader))
+                    {
+                        int colonIndex = responseHeader.IndexOf(':');
+
+                        // Skip malformed header lines that are missing the colon character.
+                        if (colonIndex > 0)
+                        {
+                            string headerName = responseHeader.Substring(0, colonIndex);
+                            string headerValue = responseHeader.Substring(colonIndex + 1);
+
+                            if (!response.Headers.TryAddWithoutValidation(headerName, headerValue))
+                            {
+                                // TODO: Skip compression related headers
+                                response.Content.Headers.TryAddWithoutValidation(headerName, headerValue);
+                            }
+                        }
+                    }
+                    return totalSize;
+                });
+        }
+
+        private unsafe static ulong CurlReceiveBodyCallback(
+            byte* buffer,
+            ulong size,
+            ulong nitems,
+            IntPtr context)
+        {
+            size *= nitems;
+            GCHandle gch = GCHandle.FromIntPtr(context);
+            var state = (RequestCompletionSource)gch.Target;
+
+            try
+            {
+                // Set task completion after all headers have been received
+                if (!state.Task.IsCompleted && !state.Task.IsCanceled && !state.Task.IsFaulted)
```

Comment:
I realize the naming's a bit confusing (it was inherited from IAsyncResult) but IsCompleted means the task is done, in one of the three final states.  Could have completed successfully, could have been canceled, could have faulted.  So the IsCanceled/IsFaulted checks are superfluous.

For that matter, you don't really need to check IsCompleted either... you can just call TrySetResult.